paw/__init__.py

```diff
@@ -0,0 +1,193 @@
+        if ord(instr) == 0 or ord(instr) == 255:
+            p = 'b'
+            print('warning: bad char detected in input')
+            self.wcount += 1
+        return p
+
+    def from_passwords(self):
+        '''
+        Reads passwords from file and generates pattern
+        '''
+        with open(self.infile, 'r', encoding='utf-8') as f:
+            for line in f:
+                self.gen_pattern(line.strip('\n'))
+        print('')
+        for key, value in self.patterns.items():
+            print('length: %d\t pattern: %s' % (key, ''.join(value)))
+
+    def gen_custom_charset(self):
+        '''
+        Generate custom charset from file or predefined charsets
+        '''
+        with open(self.infile, 'r', encoding='utf-8') as f:
+            for i, line in enumerate(f):
+                self.cset[i] = list(sorted(set(line.strip('\n'))))
+                for j in self.cset[i]:
+                    try:
+                        self.patterns[i] = ''.join(
+                            sorted(set(self.patterns[i]
+                                       + self.cset_lookup(j))))
+                    except KeyError:
+                        self.patterns[i] = self.cset_lookup(j)
+
+    def gen_hcat_cmd(self):
+        '''
+        Print hashcat string for standard patterns
+        '''
+        catstr = '-a 3 '
+        print('')
+        for k, i in self.patterns.items():
+            pattern = ''
+            hexu = ''
+            hexl = ''
+            if len(''.join(i)) > 0:
+                if '%dh' in ''.join(sorted(i)):
+                    hexu = '-1 %s%s ' % (csets['d'], csets['h'])
+                elif '%h' in ''.join(sorted(i)):
+                    hexu = '-1 %s ' % csets['h']
+                if '%di' in ''.join(sorted(i)):
+                    hexl = '-2 %s%s ' % (csets['d'], csets['i'])
+                elif '%i' in ''.join(sorted(i)):
+                    hexl = '-2 %s ' % csets['i']
+
+                for j in i:
+                    pattern += ('?' + j
+                                .replace('%', '')
+                                .replace('dh', 'h')
+                                .replace('di', 'i')
+                                .replace('h', '1')
+                                .replace('i', '2'))
+                self.catstrs[k] = (catstr + hexu + hexl +
+                                   pattern.replace('??', '?'))
+            else:
+                print('warning: pattern %d is empty' % k)
+                self.wcount += 1
+
+        for i in self.catstrs.values():
+            print(i)
+
+    def gen_pattern(self, instr):
+        '''
+        Generate pattern and position specific charset from string
+        '''
+        # Get pattern for each string position
+        pattern = ['%' + self.cset_lookup(i) for i in instr]
+
+        # Sort pattern and keep only unique values
+        pattern = [''.join(sorted(set(i))) for i in pattern]
+        for i in range(len(instr)):
```

Comment:
Indeed, though the first time was just to create the pattern, this one works on the charset that will be used, unless I am missing something

Reply:
Yeah, the question is more, why is your algorithm walking over the same data repeatedly when it could walk over it once.

Reply:
Oh, well. The first time is to create the pattern from the input string, the second time reads the values again to build the character set. I guess it would be better to move the pattern creation into the for loop and do both things at the same time.